lantiq: xway: kernel: enable SMP support

Kistelini opened the pull request:
This is a continuation of https://github.com/openwrt/openwrt/pull/3789.

I don't know what the DTS change does but it works.

Commit:
lantiq: xway: ar9: add ICU1 (2nd core for SMP)

Fixes leftover TODO from commit 6bf179b27004eb76df3e466bd080fc5a83ccf0dd

Signed-off-by: Christian Buschau <christian.buschau@mailbox.org>
Acked-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>

timocapa commented:
Works on my FB7330. Will run a WiFi bench later, before it hit 100% usage.
![image](https://user-images.githubusercontent.com/22844613/109404288-e3c59780-7964-11eb-9180-924f7fae10ee.png)

hauke commented on target/linux/lantiq/config-5.4:
```diff
@@ -1,23 +1,9 @@
 CONFIG_ARCH_32BIT_OFF_T=y
 CONFIG_ARCH_CLOCKSOURCE_DATA=y
-CONFIG_ARCH_HAS_DMA_COHERENT_TO_PFN=y
```
Which sub target was selected when you did the kernel_oldconfig?

hauke commented on target/linux/lantiq/xway/config-5.4:
```diff
@@ -43,3 +43,6 @@ CONFIG_UBIFS_FS_LZO=y
 CONFIG_USB=y
```
Please run: ```make kernel_oldconfig CONFIG_TARGET=subtarget```

Kistelini replied:
That was what I was looking for, thanks!

hauke commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/ar9.dtsi:
```diff
@@ -44,8 +44,8 @@
-			/* TODO: AR9 should have ICU1 (like VR9) too */
-			reg = <0x80200 0xc8>;
+			reg = <0x80200 0xc8	/* icu0 */
+				0x80300 0xc8>;	/* icu1 */
```
This looks good to me, if there are two VPEs there should aslo be two ICU blocks.

hauke replied:
Is there any ar9 device which uses the voice FW? This firmware should be loaded into the 2. VPE.

Kistelini replied:
The FXS ports on all ar9 devices are unsupported and there doesn't seem to be any development towards a driver and firmware support, so I thought it would be more useful to use the second VPE for OpenWRT, too.

mkresin commented:
I'm kind of concerned about side effects to the danube boards. ar9 and danube boards are sharing the same kernel configuration (xway).

Kistelini commented:
> I'm kind of concerned about side effects to the danube boards. ar9 and danube boards are sharing the same kernel configuration (xway).

Seems to work, tested on o2 Box 4421:
```
[    0.087426] smp: Bringing up secondary CPUs ...
[    0.091959] smp: Brought up 1 node, 1 CPU
```

everloop2 reviewed (approved):
https://github.com/openwrt/openwrt/pull/3789/commits/5249c32122a0809f4829b6461d8e35b823c5e49d -> changed: `/openwrt/target/linux/lantiq/files/arch/mips/boot/dts/lantiq/ar9.dtsi`.

run tested FritzBox 7320 single core and SMP enabled kernel.

everloop2 reviewed (approved):
https://github.com/openwrt/openwrt/pull/3789/commits/5249c32122a0809f4829b6461d8e35b823c5e49d -> changed: `/openwrt/target/linux/lantiq/files/arch/mips/boot/dts/lantiq/ar9.dtsi`,

run tested FritzBox 7320 single core and SMP enabled kernel.

Edit: did not check target/linux/lantiq/xway/config-5.4 as viewed.

alaviss commented:
Confirmed working with Buffalo WBMR-HP-G300H

```
# cat /proc/cpuinfo
system type		: AR9 rev 1.2
machine			: Buffalo WBMR-HP-G300H
processor		: 0
cpu model		: MIPS 34Kc V4.12
BogoMIPS		: 221.18
wait instruction	: yes
microsecond timers	: yes
tlb_entries		: 16
extra interrupt vector	: yes
hardware watchpoint	: yes, count: 4, address/irw mask: [0x0ffc, 0x0ffc, 0x0ffb, 0x0ffb]
isa			: mips1 mips2 mips32r1 mips32r2
ASEs implemented	: mips16 dsp mt
Options implemented	: tlb 4kex 4k_cache 32fpr prefetch mcheck ejtag llsc vint perf_cntr_intr_bit perf
shadow register sets	: 1
kscratch registers	: 0
package			: 0
core			: 0
VPE			: 0
VCED exceptions		: not available
VCEI exceptions		: not available

processor		: 1
cpu model		: MIPS 34Kc V4.12
BogoMIPS		: 222.41
wait instruction	: yes
microsecond timers	: yes
tlb_entries		: 16
extra interrupt vector	: yes
hardware watchpoint	: yes, count: 4, address/irw mask: [0x0ffc, 0x0ffc, 0x0ffb, 0x0ffb]
isa			: mips1 mips2 mips32r1 mips32r2
ASEs implemented	: mips16 dsp mt
Options implemented	: tlb 4kex 4k_cache 32fpr prefetch mcheck ejtag llsc vint perf_cntr_intr_bit perf
shadow register sets	: 1
kscratch registers	: 0
package			: 0
core			: 0
VPE			: 1
VCED exceptions		: not available
VCEI exceptions		: not available
```

alaviss commented:
Apparently using `irqbalance` and maxing out the router will deadlock it.

adschm commented:
Just ran kernel_oldconfig myself and received the following diff:

```
--- a/target/linux/lantiq/xway/config-5.4
+++ b/target/linux/lantiq/xway/config-5.4
@@ -4,8 +4,11 @@ CONFIG_AT803X_PHY=y
 CONFIG_BLK_MQ_PCI=y
 CONFIG_CRC16=y
 CONFIG_CRYPTO_ACOMP2=y
+CONFIG_CRYPTO_DEFLATE=y
 CONFIG_CRYPTO_HASH_INFO=y
+CONFIG_CRYPTO_LIB_POLY1305_RSIZE=2
 CONFIG_CRYPTO_LZO=y
+CONFIG_CRYPTO_ZSTD=y
 CONFIG_GENERIC_ALLOCATOR=y
 CONFIG_INPUT=y
 CONFIG_INPUT_EVDEV=y
@@ -14,6 +17,7 @@ CONFIG_INPUT_POLLDEV=y
 CONFIG_LANTIQ_ETOP=y
 CONFIG_LZO_COMPRESS=y
 CONFIG_LZO_DECOMPRESS=y
+CONFIG_MIPS_CBPF_JIT=y
 CONFIG_MTD_NAND_CORE=y
 CONFIG_MTD_NAND_ECC_SW_HAMMING=y
 CONFIG_MTD_NAND_XWAY=y
@@ -41,3 +45,8 @@ CONFIG_UBIFS_FS=y
 CONFIG_USB=y
 CONFIG_USB_COMMON=y
 CONFIG_USB_SUPPORT=y
+CONFIG_XXHASH=y
+CONFIG_ZLIB_DEFLATE=y
+CONFIG_ZLIB_INFLATE=y
+CONFIG_ZSTD_COMPRESS=y
+CONFIG_ZSTD_DECOMPRESS=y
```

Any idea why this is different?

Kistelini commented:
> Just ran kernel_oldconfig myself and received the following diff:
> ...

I don't know, I ran it again and now it's the same as yours.


> Apparently ~using `irqbalance` and~ maxing out the router CPU will deadlock it.

Actually I did not think of testing that, but I ran `stress-ng --cpu 2` on my Fritzbox 7312 and 7320 for more than an hour and there were no crashes. Could you try that too to narrow down if the crashes are actually CPU related?

alaviss commented:
> Could you try that too to narrow down if the crashes are actually CPU related?

I'll try when I have some time off but here are some more things that I noticed with my earlier usage which hopefully will let you reproduce what I saw:

- Luci ridiculously slow, it would frequently timeout.
- Maximizing the wifi + ethernet via a bufferbloat test (dslreports.com) will deadlock the system (might need 2-3 tests).
- `unbound` and `stubby` is stuck at 100% CPU usage and DNS resolution become super slow, also has a potential to lock the router up.

Kistelini commented:
I didn't try unbound and stubby but I can't reproduce the other two issues either.

alaviss commented:
> Could you try that too to narrow down if the crashes are actually CPU related?

I tried `stress-ng` but it doesn't seem to crash the router, though it looks like the CPU perf doesn't go up at all, if anything it might have gone down (see bogo ops/s).

Single core:

```
root@Buffalo:~# stress-ng -t 30 --metrics-brief --cpu 0 --cpu-method all
stress-ng: info:  [4761] setting to a 30 second run per stressor
stress-ng: info:  [4761] dispatching hogs: 1 cpu
stress-ng: info:  [4761] successful run completed in 37.70s
stress-ng: info:  [4761] stressor       bogo ops real time  usr time  sys time   bogo ops/s     bogo ops/s
stress-ng: info:  [4761]                           (secs)    (secs)    (secs)   (real time) (usr+sys time)
stress-ng: info:  [4761] cpu                  74     37.70     21.26      0.43         1.96           3.41
root@Buffalo:~# stress-ng -t 30 --metrics-brief --cpu 0 --cpu-method all
stress-ng: info:  [4915] setting to a 30 second run per stressor
stress-ng: info:  [4915] dispatching hogs: 1 cpu
stress-ng: info:  [4915] successful run completed in 30.01s
stress-ng: info:  [4915] stressor       bogo ops real time  usr time  sys time   bogo ops/s     bogo ops/s
stress-ng: info:  [4915]                           (secs)    (secs)    (secs)   (real time) (usr+sys time)
stress-ng: info:  [4915] cpu                 106     30.01     25.77      0.30         3.53           4.07
root@Buffalo:~# stress-ng -t 30 --metrics-brief --cpu 0 --cpu-method all
stress-ng: info:  [4991] setting to a 30 second run per stressor
stress-ng: info:  [4991] dispatching hogs: 1 cpu
stress-ng: info:  [4991] successful run completed in 31.12s
stress-ng: info:  [4991] stressor       bogo ops real time  usr time  sys time   bogo ops/s     bogo ops/s
stress-ng: info:  [4991]                           (secs)    (secs)    (secs)   (real time) (usr+sys time)
stress-ng: info:  [4991] cpu                 116     31.11     28.11      0.35         3.73           4.08
```

Dual core:

```
root@Buffalo:~# stress-ng -t 30 --metrics-brief --cpu 0 --cpu-method all
stress-ng: info:  [5146] setting to a 30 second run per stressor
stress-ng: info:  [5146] dispatching hogs: 2 cpu
stress-ng: info:  [5146] successful run completed in 41.83s
stress-ng: info:  [5146] stressor       bogo ops real time  usr time  sys time   bogo ops/s     bogo ops/s
stress-ng: info:  [5146]                           (secs)    (secs)    (secs)   (real time) (usr+sys time)
stress-ng: info:  [5146] cpu                 148     41.59     72.95      0.09         3.56           2.03
root@Buffalo:~# stress-ng -t 30 --metrics-brief --cpu 0 --cpu-method all
stress-ng: info:  [5454] setting to a 30 second run per stressor
stress-ng: info:  [5454] dispatching hogs: 2 cpu
stress-ng: info:  [5454] successful run completed in 41.81s
stress-ng: info:  [5454] stressor       bogo ops real time  usr time  sys time   bogo ops/s     bogo ops/s
stress-ng: info:  [5454]                           (secs)    (secs)    (secs)   (real time) (usr+sys time)
stress-ng: info:  [5454] cpu                 148     41.70     73.12      0.17         3.55           2.02
root@Buffalo:~# stress-ng -t 30 --metrics-brief --cpu 0 --cpu-method all
stress-ng: info:  [5457] setting to a 30 second run per stressor
stress-ng: info:  [5457] dispatching hogs: 2 cpu
stress-ng: info:  [5457] successful run completed in 40.19s
stress-ng: info:  [5457] stressor       bogo ops real time  usr time  sys time   bogo ops/s     bogo ops/s
stress-ng: info:  [5457]                           (secs)    (secs)    (secs)   (real time) (usr+sys time)
stress-ng: info:  [5457] cpu                 148     39.98     72.86      0.06         3.70           2.03
```

My issues with LuCI lagging is gone too, so I'm gonna give it a long run and see if any issues pops up.

argonym commented:
@abajk: Is your patch "[MIPS: smp-mt: enable all hardware interrupts on second VPE](https://patchwork.kernel.org/project/linux-mips/patch/20210606181525.761333-2-olek2@wp.pl/)" also needed to make interrupts on the second AR9 core work?

abajk commented:
> 
> 
> @abajk: Is your patch "[MIPS: smp-mt: enable all hardware interrupts on second VPE](https://patchwork.kernel.org/project/linux-mips/patch/20210606181525.761333-2-olek2@wp.pl/)" also needed to make interrupts on the second AR9 core work?

Without this patch, only interrupts connected to the IP6 and IP7 lines should work on the second VPE.  On the xRX200 they work for pcie (144 ICU inteerupt connected to IP6 ) and not work for ethernet (72 and 73 lines connected to IP2-5).

alaviss commented:
One month report: It's rock solid.

The issue with my router hanging seems to boils down to OOM, I reconfigured the system to use less memory (unbound was triggering OOM killer a lot) and added zram which solved the issue.

Commit:
lantiq: xway: refresh kernel config

Run of 'make kernel_oldconfig CONFIG_TARGET=subtarget'

Signed-off-by: Christian Buschau <christian.buschau@mailbox.org>
Acked-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>

Commit:
lantiq: xway: kernel: enable SMP support

Enabled CONFIG_MIPS_MT_SMP and CONFIG_SCHED_SMT.

Tested on FRITZ!Box 7330 SL, 7312 and o2 Box 4421.

Signed-off-by: Christian Buschau <christian.buschau@mailbox.org>
Acked-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>

Kistelini commented:
I did some testing and this change nearly doubles WireGuard performance. Since https://github.com/openwrt/openwrt/commit/fbd33d61648ae8982fbada7ad3b6d8222b367ab5 there shouldn't be anything left, can someone have a look at this again?

ventod commented:
Just wanted to add my experience. Running this on my BT Home Hub 5A since 2022-01-18 aka: when I last compiled my personal image (a really bare-bone image TBH but at least is stable as hell and I've never had a crash). I manually moved 2 IRQs to the other cpu (20 = ath9k and 72 = xrx200_net_rx).

IIRC older kernel automatically balanced IRQs between available CPUs but not this does not work anymore?

Kistelini commented:
@ventod Your device is part of the xrx200 subtarget and this pull request changes things for xway so it should be irrelevant for your device.

I'd really like to get this merged. Maybe @xdarklight can say something because lantiq stuff?

abajk commented:
AR9 is similar to VR9. It has a 34Kc core, a built-in 3 port switch. It may be better in the long-term to move these devices to the xrx200 target. This may happen during the conversion to DSA.

xdarklight commented:
Personally I'm fine with merging this but I don't have time to test it.
The only concern I had is the same that @mkresin also mentioned: impact on Danube SoCs. My understanding is that this has been tested. If @abajk also sees no reason to NACK this then I suggest that @hauke should merge it.

```Acked-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>```

djStolen commented:
Hi all, 

I have a bunch of O2 6431 and am really interested in this improvement.

I would gladly test it, just need couple of clarifications.

1) Can I patch it on 22.03.2?
2) Can I build a default config with just this patch on top?
3) Or. Are those 2 lines the only once that have to be patched?

Thanks in advance

Kistelini commented:
@djStolen Your device is xrx200 which already supports SMP. This patch is for xway.

djStolen commented:
> @djStolen Your device is xrx200 which already supports SMP. This patch is for xway.

Thanks for your reply.

How come with the pre-built binary of 22.03.2 I only see the load on 1 core in htop?

What am I missing?

hauke commented:
Please add the kernel configuration changes also for kernel 5.15. 
Please also add the `Acked-by`tags you got to the commits too. 

Did someone test this on Danube, I have seen multiple people testing his on AR9?

Kistelini commented:
@hauke done. I tested it on a [o2 Box 4421](https://github.com/openwrt/openwrt/pull/3946#issuecomment-787544406) but I don't have that device anymore.

Kistelini commented:
> > @djStolen Your device is xrx200 which already supports SMP. This patch is for xway.
> 
> Thanks for your reply.
> 
> How come with the pre-built binary of 22.03.2 I only see the load on 1 core in htop?
> 
> What am I missing?

https://github.com/openwrt/openwrt/blob/fdcd383919e9ba6a8f6558a1eff9abf7175e872e/target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_arcadyan_vgv7510kw22.dtsi#L10
The `nosmp` option. I don't know why it's set.

djStolen commented:
> > > @djStolen Your device is xrx200 which already supports SMP. This patch is for xway.
> > 
> > Thanks for your reply.
> > 
> > How come with the pre-built binary of 22.03.2 I only see the load on 1 core in htop?
> > 
> > What am I missing?
> 
> https://github.com/openwrt/openwrt/blob/fdcd383919e9ba6a8f6558a1eff9abf7175e872e/target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_arcadyan_vgv7510kw22.dtsi#L10
> The `nosmp` option. I don't know why it's set.

Strange, if it supports. Will have a look into it, run a compile with removing only that switch.

Thanks

hauke commented:
> > > @djStolen Your device is xrx200 which already supports SMP. This patch is for xway.
> > 
> > 
> > Thanks for your reply.
> > How come with the pre-built binary of 22.03.2 I only see the load on 1 core in htop?
> > What am I missing?
> 
> https://github.com/openwrt/openwrt/blob/fdcd383919e9ba6a8f6558a1eff9abf7175e872e/target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_arcadyan_vgv7510kw22.dtsi#L10
> 
> The `nosmp` option. I don't know why it's set.

`nosmp` i set here to allow loading the voice firmware onto the second VPE instead of using it for Linux SMP.

djStolen commented:
> > > > @djStolen Your device is xrx200 which already supports SMP. This patch is for xway.
> > > 
> > > 
> > > Thanks for your reply.
> > > How come with the pre-built binary of 22.03.2 I only see the load on 1 core in htop?
> > > What am I missing?
> > 
> > https://github.com/openwrt/openwrt/blob/fdcd383919e9ba6a8f6558a1eff9abf7175e872e/target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_arcadyan_vgv7510kw22.dtsi#L10
> > 
> > The `nosmp` option. I don't know why it's set.
> 
> `nosmp` i set here to allow loading the voice firmware onto the second VPE instead of using it for Linux SMP.

Aaaa, ok, got it. I use my device without phone, so don't really need it. I quests the functionality is anyhow not used often anymore, so might be interesting for broader audience to have the extra core (e.g. for smarthome server)

hauke commented:
Thank you for the patches, I applied them to master in fb0256e88e2f495d08cd4c45eb243365a80696f1, d3527bb83be949d2f4bc4431d5750a0fadc6c1f5 and 084c20f6c54180f2dc6b3efc1c5ba9b62afbaf26.
